x-pack/filebeat/input/httpjson/config_okta_auth.go

```diff
@@ -0,0 +1,192 @@
+		return nil, fmt.Errorf("oauth2 client: error generating Okta JWT: %w", err)
+	}
+
+	token, err := exchangeForBearerToken(ctx, oktaJWT, conf)
+	if err != nil {
+		return nil, fmt.Errorf("oauth2 client: error exchanging Okta JWT for bearer token: %w", err)
+	}
+
+	tokenSource := &customTokenSource{
+		conf:    conf,
+		ctx:     ctx,
+		oktaJWK: oktaJWK,
+		token: &oauth2.Token{
+			AccessToken: token.AccessToken,
+		},
+	}
+	// reuse the tokenSource to refresh the token (automatically calls the custom Token() method when token is no longer valid).
+	client := oauth2.NewClient(ctx, oauth2.ReuseTokenSource(token, tokenSource))
+
+	return client, nil
+}
+
+// Token implements the oauth2.TokenSource interface and helps to implement custom token refresh logic.
+// parent context is passed via the customTokenSource struct since we cannot modify the function signature here.
+func (ts *customTokenSource) Token() (*oauth2.Token, error) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	oktaJWT, err := generateOktaJWT(ts.oktaJWK, ts.conf)
+	if err != nil {
+		return nil, fmt.Errorf("error generating Okta JWT: %w", err)
+	}
+	token, err := exchangeForBearerToken(ts.ctx, oktaJWT, ts.conf)
+	if err != nil {
+		return nil, fmt.Errorf("error exchanging Okta JWT for bearer token: %w", err)
+
+	}
+
+	return token, nil
+}
+
+func generateOktaJWT(oktaJWK string, cnf *oauth2.Config) (string, error) {
+	// unmarshal the JWK into a map
+	var jwkData map[string]string
+	err := json.Unmarshal([]byte(oktaJWK), &jwkData)
+	if err != nil {
+		return "", fmt.Errorf("error decoding JWK: %w", err)
+	}
+
+	// create a RSA private key from JWK components
+	decodeBase64 := func(key string) (*big.Int, error) {
+		data, err := base64.RawURLEncoding.DecodeString(jwkData[key])
+		if err != nil {
+			return nil, fmt.Errorf("error decoding RSA JWK component %s: %w", key, err)
+		}
+		return new(big.Int).SetBytes(data), nil
+	}
+
+	n, err := decodeBase64("n")
+	if err != nil {
+		return "", err
+	}
+	e, err := decodeBase64("e")
+	if err != nil {
+		return "", err
+	}
+	d, err := decodeBase64("d")
+	if err != nil {
+		return "", err
+	}
+	p, err := decodeBase64("p")
+	if err != nil {
+		return "", err
+	}
+	q, err := decodeBase64("q")
+	if err != nil {
+		return "", err
+	}
+	dp, err := decodeBase64("dp")
+	if err != nil {
+		return "", err
+	}
+	dq, err := decodeBase64("dq")
+	if err != nil {
+		return "", err
+	}
+	qi, err := decodeBase64("qi")
+	if err != nil {
+		return "", err
+	}
+
+	privateKeyRSA := &rsa.PrivateKey{
+		PublicKey: rsa.PublicKey{
+			N: n,
+			E: int(e.Int64()),
+		},
+		D:      d,
+		Primes: []*big.Int{p, q},
+		Precomputed: rsa.PrecomputedValues{
+			Dp:   dp,
+			Dq:   dq,
+			Qinv: qi,
+		},
+	}
+
+	// create a JWT token using required claims and sign it with the private key
+	tok, err := jwt.NewBuilder().Audience([]string{cnf.Endpoint.TokenURL}).
+		Issuer(cnf.ClientID).
+		Subject(cnf.ClientID).
+		IssuedAt(time.Now()).
+		Expiration(time.Now().Add(time.Hour)).
```

Comment:
Shouldn't these two "now" values be exactly the same? That is, store a `now := time.Now()` and then use `IssuedAt(now)` and `Expiration(now.Add(time.Hour))`.